stylix/mk-target.nix

```diff
@@ -0,0 +1,102 @@
+{
+  name,
+  humanName,
+  autoEnable ? true,
+  extraOptions ? { },
+  configElements ? [ ],
+  generalConfig ? null,
+}:
+{ config, lib, ... }:
+let
+  cfg = config.stylix.targets.${name};
+in
+{
+  options.stylix.targets.${name} = lib.attrsets.unionOfDisjoint {
+    enable = config.lib.stylix.mkEnableTarget humanName autoEnable;
+  } extraOptions;
+
+  config =
+    let
+      args =
+        fn:
+        let
+          args = lib.functionArgs fn;
+          invalidArgs = lib.filterAttrs (_: value: value) args;
+          invalidArgsString = builtins.concatStringsSep ", " (
+            builtins.attrNames invalidArgs
+          );
+        in
+        lib.throwIf (invalidArgs != { })
+          "function arguments should not have default values: ${invalidArgsString}"
+          (builtins.attrNames args);
+      provideStylixArgs =
+        args':
+        (builtins.listToAttrs (
+          map (
+            arg:
+            if arg == "cfg" then
+              {
+                name = "cfg";
+                value = cfg;
+              }
+            else if arg == "colors" then
+              {
+                name = arg;
+                value = config.lib.stylix.${arg};
+              }
+            else
+              {
+                name = arg;
+                value = config.stylix.${arg};
+              }
+          ) args'
+        ));
+      provideAvailableStylixArgs =
+        args':
+        (builtins.listToAttrs (
+          map (
+            arg:
+            if arg == "cfg" then
+              {
+                name = "cfg";
+                value = cfg;
+              }
+            else if arg == "colors" then
+              {
+                name = arg;
+                value = config.lib.stylix.${arg};
+              }
+            else if config.stylix ? ${arg} then
+              {
+                name = arg;
+                value = config.stylix.${arg};
+              }
+            else
+              throw "stylix: Stylix arg was requested by mkTarget but does not exist: ${arg}"
+          ) args'
+        ));
+      mkConfig = fn: fn (provideAvailableStylixArgs (args fn));
+      mkConditionalConfig =
+        fn:
+        let
+          args' = args fn;
+        in
+        lib.mkIf (builtins.all (
+          arg:
+          if arg == "cfg" then
+            true
+          else if arg == "colors" then
+            config.lib.stylix.${arg} != null
+          else if config.stylix ? ${arg} then
+            config.stylix.${arg} != null
+          else
+            throw "stylix: Stylix arg was requested by mkTarget but does not exist: ${arg}"
+        ) args') (fn (provideStylixArgs args'));
+    in
+    lib.mkIf (config.stylix.enable && cfg.enable) (
+      lib.mkMerge (
+        lib.optional (generalConfig != null) (mkConfig generalConfig)
+        ++ (map mkConditionalConfig configElements)
```

Comment:
Maybe some comments would help remind us at this point what `generalConfig` and `configElements` are and/or why we're using them as config definitions?